Remove paket

bender2k14 opened the pull request:
(Will eventually...) Fix #206.

The first commit removes all references to paket that I could find and adds the typical type of NuGet package references using the versions that were given in the `paket.dependencies`.

However, two tests fail for me in this commit.  This commit is essentially what prompted me to create issue #201 about the possibility of a test failing in `master`.  @moodmosaic, what is the result when you run theses tests?

The second commit is a workaround I found to get the tests to pass.

I am creating this PR as a draft because there might be a better way to get the tests to pass than the workaround that presented here.  I also haven't run the benchmark tests.  I was planning to run them after the tests pass for everyone and a consensus has been reached about the best way to get the test to pass.

moodmosaic commented on tests/Hedgehog.Benchmarks/Hedgehog.Benchmarks.fsproj:
```diff
@@ -2,20 +2,23 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Could that be the reason why those tests fail on your local?

bender2k14 replied:
I don't know what you mean.  Can you elaborate?

This specific line is in the Benchmark project.  It is changed only to match the same change in the main test project.  That change was part of a workaround I found that gets the tests to _pass_.  Before I made that change, two tests failed.  To repeat, two tests fail for me on the first commit and all tests pass for me on the second commit.

moodmosaic replied:
Nah, you're right. I didn't notice it was on the benchmarks project. 👍

moodmosaic commented on src/Hedgehog/Hedgehog.fsproj:
```diff
@@ -16,9 +16,9 @@
   </ItemGroup>
-  <Import Project="..\..\.paket\Paket.Restore.targets" />
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="4.1.17" />
```
What does the `Update` element means? Is it the same as e.g.

```xml
<PackageReference Include="FSharp.Core" Version="[4.1.17,)" />
```

?


(We want to depend on FSharp.Core 4.1.17, the earliest one that supports .NET standard, but not require that exact version in the generated NuGet Package, just require it as a minimum bound.)

bender2k14 replied:
`[4.1.17,)` doesn't work because `FSharp.Core` version `4.7.0` and higher doesn't support .NET Standard 1.6.  That is another reason why adding support for .NET Standard 2.0 (c.f. issue #204) is a good idea.

moodmosaic replied:
OK, good to know that.

bender2k14 replied:
I don't know what the difference is between `Include` and `Update`.  I used `Update` because that is what Visual Studio created when I added `FSharp.Core` as a dependency.  If I change it to `Include`, then I can build through Visual Studio, but running `dotnet build` for that project fails.
```
twilliams@EDMONDS MINGW64 /c/Code/ThirdParty/fsharp-hedgehog (remove_paket)
$ dotnet build src/Hedgehog/Hedgehog.fsproj
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202: Package FSharp.Core 4.7.1 is not compatible with netstandard1.6 (.NETStandard,Version=v1.6). Package FSharp.Core 4.7.1 supports:
C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202:   - net45 (.NETFramework,Version=v4.5)
C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202:   - netstandard2.0 (.NETStandard,Version=v2.0)
  Failed to restore C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj (in 613 ms).

Build FAILED.

C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202: Package FSharp.Core 4.7.1 is not compatible with netstandard1.6 (.NETStandard,Version=v1.6). Package FSharp.Core 4.7.1 supports:
C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202:   - net45 (.NETFramework,Version=v4.5)
C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj : error NU1202:   - netstandard2.0 (.NETStandard,Version=v2.0)
    0 Warning(s)
    1 Error(s)
```

In contrast, `dotnet build` for that project leaving it as `Update` gives
```
twilliams@EDMONDS MINGW64 /c/Code/ThirdParty/fsharp-hedgehog (remove_paket)
$ dotnet build src/Hedgehog/Hedgehog.fsproj
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj (in 686 ms).
  Hedgehog -> C:\Code\ThirdParty\fsharp-hedgehog\src\Hedgehog\bin\Debug\netstandard1.6\Hedgehog.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)
```

It isn't quite fair of me to post these runs without also providing exact commits.  I can back to this with an update when my commits have settled down.

moodmosaic replied:
Do whatever works better for you :+1: My understanding is that with `Update` you force a specific version solution-wide. I just can't seem to find anything in the docs...

bender2k14 replied:
> `[4.1.17,)` doesn't work because `FSharp.Core` version `4.7.0` and higher doesn't support .NET Standard 1.6. That is another reason why adding support for .NET Standard 2.0 (c.f. issue #204) is a good idea.

In hindsight, not sure why I said that.  Using `[4.1.17,)` is working in the current branch.

This conversation continued in the main thread.  I will resolve this "conversation".

moodmosaic commented on tests/Hedgehog.Tests/Hedgehog.Tests.fsproj:
```diff
@@ -12,10 +12,22 @@
   </ItemGroup>
-  <Import Project="..\..\.paket\Paket.Restore.targets" />
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="4.2.3" />
```
See comment above about `Update`. And perhaps we want `Version="4.*"`?

bender2k14 replied:
`4.*` probably isn't right because `FSharp.Core` version `4.7.0` doesn't support .NET Core 2.0 but version `4.6.2` does

bender2k14 replied:
It is ok to depend on a specific version `FSharp.Core` here because this project isn't a library project consumed by others but an executable project only used "internally" by the "Hedgehog developers" for testing.  

paket's lock file shows that it resolved version 4.6.1 for the `Test` group.  I picked version 4.2.3 because it matches the constraint imposed by Unqoute before I thought to look at the lock file.  I think picking any of these versions is fine.
https://github.com/hedgehogqa/fsharp-hedgehog/blob/2a68ee22e85b5d9026b3b21a6a0a04bce9cef086/paket.lock#L475

moodmosaic commented on tests/Hedgehog.Benchmarks/Hedgehog.Benchmarks.fsproj:
```diff
@@ -2,20 +2,23 @@
-  <Import Project="..\..\.paket\Paket.Restore.targets" />
+  
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="4.2.3" />
```
See comment above about `Update`. And perhaps we want `Version="4.*"`?

bender2k14 replied:
It is ok to depend on a specific version `FSharp.Core` here because this project isn't a library project consumed by others but an executable project only used "internally" by the "Hedgehog developers" for testing.  

paket's lock file shows that it resolved version 4.6.1 for the `Test` group.  I picked version 4.2.3 because it matches the constraint imposed by Unqoute before I thought to look at the lock file.  I think picking any of these versions is fine.
https://github.com/hedgehogqa/fsharp-hedgehog/blob/2a68ee22e85b5d9026b3b21a6a0a04bce9cef086/paket.lock#L475

bender2k14 commented:
> Left some small comments in a first review round. 👍

Those are good comments.  Before I address them, I would prefer get on the same page regarding the test results.

Two tests fail for me on the second (and last) commit.  This is essentially what prompted me to create issue #201 about the possibility of a test failing in master. @moodmosaic, what is the result when you run theses tests?

moodmosaic commented:
I just checked out the branch and all tests are passing:

```
git stash && git fetch -p && git checkout master
git clean -dxf && git checkout .
git checkout -b bender2k14-remove_paket master
git pull https://github.com/bender2k14/fsharp-hedgehog.git remove_paket
dotnet test -v n tests/Hedgehog.Tests/Hedgehog.Tests.fsproj
```

Output:
```
Build started 7/7/2020 07:19:04 AM.
     1>Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" on node 1 (Restore target(s)).
     1>_GetAllRestoreProjectPathItems:
         Determining projects to restore...
       Restore:
         Committing restore...
         Committing restore...
         Assets file has not changed. Skipping assets file writing. Path: ..\fsharp-hedgehog\tests\Hedgehog.Tests\obj\project.assets.json
         Assets file has not changed. Skipping assets file writing. Path: ..\fsharp-hedgehog\src\Hedgehog\obj\project.assets.json
         Restored ..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj (in 34 ms).
         Restored ..\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj (in 34 ms).
         
         NuGet Config files used:
             ..\NuGet\NuGet.Config
         
         Feeds used:
             https://api.nuget.org/v3/index.json
         All projects are up-to-date for restore.
     1>Done Building Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (Restore target(s)).
   1:7>Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" on node 1 (VSTest target(s)).
     1>BuildProject:
         Build started, please wait...
   1:7>Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (1:7) is building "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (1:8) on node 1 (default targets).
   1:8>Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (1:8) is building "..\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj" (2:6) on node 1 (default targets).
     2>GenerateTargetFrameworkMonikerAttribute:
       Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the input files.
       CoreGenerateAssemblyInfo:
       Skipping target "CoreGenerateAssemblyInfo" because all output files are up-to-date with respect to the input files.
       CoreCompile:
       Skipping target "CoreCompile" because all output files are up-to-date with respect to the input files.
       GenerateBuildDependencyFile:
       Skipping target "GenerateBuildDependencyFile" because all output files are up-to-date with respect to the input files.
       CopyFilesToOutputDirectory:
         Hedgehog -> ..\fsharp-hedgehog\src\Hedgehog\bin\Debug\netstandard1.6\Hedgehog.dll
     2>Done Building Project "..\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj" (default targets).
     1>GenerateTargetFrameworkMonikerAttribute:
       Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the input files.
       CoreGenerateAssemblyInfo:
       Skipping target "CoreGenerateAssemblyInfo" because all output files are up-to-date with respect to the input files.
       CoreCompile:
       Skipping target "CoreCompile" because all output files are up-to-date with respect to the input files.
       _CopyFilesMarkedCopyLocal:
         Touching "..\fsharp-hedgehog\tests\Hedgehog.Tests\obj\Debug\netcoreapp3.1\Hedgehog.Tests.fsproj.CopyComplete".
       _CopyOutOfDateSourceItemsToOutputDirectory:
       Skipping target "_CopyOutOfDateSourceItemsToOutputDirectory" because all output files are up-to-date with respect to the input files.
       GenerateBuildDependencyFile:
       Skipping target "GenerateBuildDependencyFile" because all output files are up-to-date with respect to the input files.
       GenerateBuildRuntimeConfigurationFiles:
       Skipping target "GenerateBuildRuntimeConfigurationFiles" because all output files are up-to-date with respect to the input files.
       CopyFilesToOutputDirectory:
         Hedgehog.Tests -> ..\fsharp-hedgehog\tests\Hedgehog.Tests\bin\Debug\netcoreapp3.1\Hedgehog.Tests.dll
     1>Done Building Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (default targets).
     1>BuildProject:
         Build completed.
         
Test run for ..\fsharp-hedgehog\tests\Hedgehog.Tests\bin\Debug\netcoreapp3.1\Hedgehog.Tests.dll(.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.6.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 3.1.5)
[xUnit.net 00:00:00.33]   Discovering: Hedgehog.Tests
[xUnit.net 00:00:00.40]   Discovered:  Hedgehog.Tests
[xUnit.net 00:00:00.40]   Starting:    Hedgehog.Tests
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 30, x: 0) [63ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 3863623312507393319, value: 7344074043290227165, gamma: 12297829382473034405) [64ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 128, x: 63) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 1024, x: 511) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 256, x: 127) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 2, x: 0) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 512, x: 255) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the bounds at(sz: 3, x: 0) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 30, y: 0) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 256, y: 127) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 1024, y: 511) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 2, y: 0) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 512, y: 255) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 3, y: 0) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant origin returns correct result<Int32>(x: 128, y: 63) [< 1ms]
  û Hedgehog.Tests.RangeTests.clamp truncates a value so it stays within some range <Int32>(x: 5, y: 10, n: 15, expected: 10) [2ms]
  û Hedgehog.Tests.RangeTests.clamp truncates a value so it stays within some range <Int32>(x: 5, y: 10, n: 0, expected: 5) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 30) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 1024) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 3) [1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 512) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 8746728143070965849, value: 600595566262245170, gamma: 12297829382473034421) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 2) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 17928928724259255865, value: 13651108307767328632, gamma: 12297829382473034413) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 13742489918233434733, value: 9920554987610416076, gamma: 12297829382473034405) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 256) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 9188376289577737681, value: 307741759840609752, gamma: 12297829382473034411) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int32 range(sz: 1) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 3484863033197266566, value: 15306535823716590088, gamma: 12297829382473034405) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 17288824720004427893, value: 12360157267739240775, gamma: 12297829382473034421) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 9675543792836633510, value: 3341781972484278810, gamma: 12297829382473034405) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 9864166656744503064, value: 1931727433621677744, gamma: 12297829382473034411) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 17584158569056203712, value: 6781260234005250507, gamma: 12297829382473034413) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 420250731748546163, value: 1471112649570176389, gamma: 12297829382473034421) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 8521984098521027180, value: 8100917074368564322, gamma: 12297829382473034421) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 11302328716527294647, value: 125750466559701114, gamma: 12297829382473034413) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 8743694738624347980, value: 8606169619657412120, gamma: 12297829382473034413) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 7046029254386353131, value: 15210016002011668638, gamma: 12297829382473034411) [< 1ms]
  û Hedgehog.Tests.SeedTests.Seed.from 'fixes' the ?-value(x: 17885559969949501885, value: 11409286845259996466, gamma: 12297829382473034411) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 1) [1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 2) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 1024) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 512) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 30) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Byte range(sz: 3) [< 1ms]
  û Hedgehog.Tests.RangeTests.exponentialBounded uses the full range of a data type - example 3 [3ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 512) [1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 30) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 1024) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 3) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 2) [< 1ms]
  û Hedgehog.Tests.RangeTests.constantBounded bounds returns correct result - Int64 range(sz: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.linearBounded uses the full range of a data type - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 2) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 3) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 30) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 512) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton origin returns correct result<Int32>(x: 1024) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 1024, x: 512) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 128, x: 64) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 30, x: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 3, x: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 256, x: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 512, x: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.singleton bounds returns correct result<Int32>(sz: 2, x: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 1024) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 2) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 512) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 3) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 30) [< 1ms]
  û Hedgehog.Tests.RangeTests.range from -x to x, with the origin at(x: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.linear scales the second bound relative to the size - example 3 [1ms]
  û Hedgehog.Tests.RangeTests.exponential scales the second bound exponentially relative to the size - example 3 [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 128, x: 63, y: 64) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 30, x: 0, y: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 2, x: 0, y: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 3, x: 0, y: 1) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 1024, x: 511, y: 512) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 512, x: 255, y: 256) [< 1ms]
  û Hedgehog.Tests.RangeTests.constant bounds returns correct result<Int32>(sz: 256, x: 127, y: 128) [< 1ms]
  û Hedgehog.Tests.RangeTests.linearBounded uses the full range of a data type - example 2 [1ms]
  û Hedgehog.Tests.RangeTests.exponentialFrom scales the bounds exponentially relative to the size - example 3 [21ms]
  û Hedgehog.Tests.RangeTests.linearFrom scales the bounds relative to the size - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.linearFrom scales the bounds relative to the size - example 3 [1ms]
  û Hedgehog.Tests.RangeTests.linear scales the second bound relative to the size - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.exponentialFrom scales the bounds exponentially relative to the size - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.linear scales the second bound relative to the size - example 2 [8ms]
  û Hedgehog.Tests.RangeTests.linearBounded uses the full range of a data type - example 3 [1ms]
  û Hedgehog.Tests.RangeTests.exponential scales the second bound exponentially relative to the size - example 2 [1ms]
  û Hedgehog.Tests.RangeTests.exponentialBounded uses the full range of a data type - example 2 [1ms]
  û Hedgehog.Tests.RangeTests.exponentialBounded uses the full range of a data type - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.exponentialFrom scales the bounds exponentially relative to the size - example 2 [< 1ms]
  û Hedgehog.Tests.RangeTests.exponential scales the second bound exponentially relative to the size - example 1 [1ms]
  û Hedgehog.Tests.RangeTests.linearFrom scales the bounds relative to the size - example 2 [1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 256, destination: 128) [14ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 128, destination: 64) [2ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 2, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 30, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 1024, destination: 512) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 512, destination: 256) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks by edging towards a destination number(x0: 3, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.removes produces all permutations of removing 'k' elements from a list - example 2 [5ms]
  û Hedgehog.Tests.ShrinkTests.towards returns empty list when run out of shrinks(x0: 1024, destination: 1024) [1ms]
  û Hedgehog.Tests.ShrinkTests.towards returns empty list when run out of shrinks(x0: 1, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards returns empty list when run out of shrinks(x0: 30, destination: 30) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble returns empty list when run out of shrinks(x0: 1024, destination: 1024) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble returns empty list when run out of shrinks(x0: 30, destination: 30) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble returns empty list when run out of shrinks(x0: 1, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 1) [1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: -4096) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: -1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 3) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 10240) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: -2048) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 2) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 0) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 8192) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 30) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 256) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: 128) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves Produces a list containing the results of halving a number(n: -8) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 128, destination: 64) [1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 1024, destination: 512) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 512, destination: 256) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 3, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 2, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 30, destination: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks by edging towards a destination number(x0: 256, destination: 128) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.removes permutes a list by removing 'k' consecutive elements from it [17ms]
  û Hedgehog.Tests.MinimalTests.greedy traversal with a predicate yields the perfect minimal shrink [108ms]
  û Hedgehog.Tests.MinimalTests.isDefaultGen returns true for default value of Gen<Exp> [< 1ms]
  û Hedgehog.Tests.MinimalTests.genExp is not its default value [< 1ms]
  û Hedgehog.Tests.MinimalTests.isDefaultGen returns false for constant Gen<Exp> Lit 0 [13ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 256) [32ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 512) [158ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 1024) [749ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 2) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 128) [20ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 30) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.elems shrinks each element in input list using a supplied shrinker(n: 3) [7ms]
  û Hedgehog.Tests.ShrinkTests.halves produces a list containing the progressive halving of an integral - example 3 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 128) [10ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 512) [19ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 30) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 1024) [210ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 3) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 2) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 1) [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list produces a smaller permutation of the input list(n: 256) [64ms]
  û Hedgehog.Tests.ShrinkTests.list shrinks a list by edging towards the empty list - example 2 [3ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks an integral number by edging towards a destination - exmaple 1 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks a floating-point number by edging towards a destination - example 2 [2ms]
  û Hedgehog.Tests.ShrinkTests.halves produces a list containing the progressive halving of an integral - example 2 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.halves produces a list containing the progressive halving of an integral - example 1 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.list shrinks a list by edging towards the empty list - example 1 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks an integral number by edging towards a destination - exmaple 3 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.removes produces all permutations of removing 'k' elements from a list - example 3 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towards shrinks an integral number by edging towards a destination - exmaple 2 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.towardsDouble shrinks a floating-point number by edging towards a destination - example 1 [< 1ms]
  û Hedgehog.Tests.ShrinkTests.removes produces all permutations of removing 'k' elements from a list - example 1 [< 1ms]
  û Hedgehog.Tests.GenTests.unicode doesn't return any surrogate [1s 401ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 16) [11ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 32) [< 1ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 256) [< 1ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 128) [< 1ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 512) [2ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 8) [< 1ms]
  û Hedgehog.Tests.GenTests.dateTime creates System.DateTime instances(count: 64) [< 1ms]
  û Hedgehog.Tests.GenTests.dateTime shrinks to correct mid-value [70ms]
  û Hedgehog.Tests.GenTests.uint64 can create exponentially bounded integer [8ms]
  û Hedgehog.Tests.GenTests.int64 can create exponentially bounded integer [5ms]
  û Hedgehog.Tests.GenTests.unicode doesn't return any noncharacter(nonchar: 65534) [725ms]
[xUnit.net 00:00:03.74]   Finished:    Hedgehog.Tests
  û Hedgehog.Tests.GenTests.unicode doesn't return any noncharacter(nonchar: 65535) [868ms]
  û Hedgehog.Tests.GenTests.dateTime randomly generates value between max and min ticks [1ms]

Test Run Successful.
Total tests: 185
     Passed: 185
 Total time: 5.0551 Seconds
     1>Done Building Project "..\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (VSTest target(s)).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:06.47

```

bender2k14 commented:
Sorry @moodmosaic.  I was unclear.  I meant to have you run the tests on the first commit of this branch.

To help improve clarity, I pushed a separate branch with that commit as its head and created PR #208 for it.  As you can see in that PR, the tests fail for the Travis CI build.

This suggests that I didn't remove paket correctly.  Something about the obtained dependencies is different, and I have not yet been able to determine what the difference is.

On the other hand, as I stated in https://github.com/hedgehogqa/fsharp-hedgehog/issues/201#issuecomment-653599642, it seems suspicious that the tests fail for precisely the reason given in the commented out warning on `genExp`.

bender2k14 commented:
I found a different [workaround](https://github.com/hedgehogqa/fsharp-hedgehog/pull/207/commits/a598c7336494a127740e154fb5a16480d7121a8c) that I like, which is to move `expGen` into the test (and then I had to move `genName` as well).

bender2k14 commented:
In Elmish.WPF, we depend on Elmish via [the line](https://github.com/elmish/Elmish.WPF/blob/02d1edba12c3e3464895e2141cfdcc4e38052d6a/src/Elmish.WPF/Elmish.WPF.fsproj#L46)...
```
<PackageReference Include="Elmish" Version="[3.0.3, 4)" />
```

...so that is some anecdotal evidence in favor of `Include`.

moodmosaic commented:
OK, I think Google-superpowers worked this time :rocket: Here's a nice [article](https://orcharddojo.net/blog/how-to-use-the-same-version-of-orchard-core-nuget-packages-in-every-project-across-my-solution-orchard-nuggets) which explains how `Update` works (and a [tweet](https://twitter.com/davkean/status/980076449636917249) as well).

moodmosaic commented on src/Hedgehog/paket.template:
```diff
@@ -1,26 +0,0 @@
-    https://github.com/hedgehogqa/fsharp-hedgehog/raw/master/img/hedgehog-logo.png
-tags
-    fsharp, f#, testing
-files
```
@ThisFunctionalTom any ideas if this is possible with `dotnet pack`?

bender2k14 replied:
> I just converted paket.template to NuGet metadata properties and pushed to your branch. (There's still the [`files`](https://github.com/hedgehogqa/fsharp-hedgehog/pull/207#pullrequestreview-445389964) field missing but that isn't yet part of any release, so it can be migrated also separately.)

Did you answer your own question here?...or is there still something I could look into?

moodmosaic replied:
I meant that we've taken care of everything except the `files` field which basically copies the sources into the NuGet Package so they can be consumed by Fable.

This was originally added by @ThisFunctionalTom in https://github.com/hedgehogqa/fsharp-hedgehog/pull/194/files#diff-5cc85c81aa919cd092437e0ef6744e25 and I think the equivalent in `dotnet pack` is the [`IncludeSource`](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#includesource) property.

It's the only thing left to do, and I'm planning to look into it soon (perhaps even today). After that we should be good.

moodmosaic commented:
> * [ ]  NuGet Package generated from `dotnet pack` is the "same" as the one generated via Paket.

I just converted paket.template to NuGet metadata properties and pushed to your branch. (There's still the [`files`](https://github.com/hedgehogqa/fsharp-hedgehog/pull/207#pullrequestreview-445389964) field missing but that isn't yet part of any release, so it can be migrated also separately.)

I've marked that one as done.

moodmosaic commented:
> * [ ]  Make sure `Hedgehog` project doesn't require a specific version of `FSharp.Core`.  Maybe use `Update` instead of `Include` in `PackageReference`.

I think this can also be marked as done, as

* the current NuGet Package (generated via `paket pack`) depends on
  * FSharp.Core (>= 4.1.17)
* the newly created NuGet Package (generated via `dotnet pack`) depends on
  * FSharp.Core (>= 4.1.17 && <= 4.6.2)

### dotnet pack (new)

![image](https://user-images.githubusercontent.com/287532/87025018-1ce4a800-c1da-11ea-8ad0-fb699a36031b.png)

### paket pack (current)

![image](https://user-images.githubusercontent.com/287532/87025209-6503ca80-c1da-11ea-85bf-83d68a7eea59.png)

bender2k14 commented:
> the newly created NuGet Package (generated via dotnet pack) depends on
> - FSharp.Core (>= 4.1.17 && <= 4.6.2)

The upper bound shouldn't be necessary.  I just created a test project where I target .NET Core 3.1, depend on the currently released Hedgehod NuGet package, and everything worked.

I will remove the upper bound.

bender2k14 commented:
> About the `Update` I came across [this](https://www.strathweb.com/2018/07/solution-wide-nuget-package-version-handling-with-msbuild-15/) article that sheds some light on.

> Here's a nice [article](https://orcharddojo.net/blog/how-to-use-the-same-version-of-orchard-core-nuget-packages-in-every-project-across-my-solution-orchard-nuggets) which explains how `Update` works (and a [tweet](https://twitter.com/davkean/status/980076449636917249) as well).

> If I change it to `Include`, then I can build through Visual Studio, but running `dotnet build` for that project fails.
> ...
> In contrast, `dotnet build` for that project leaving it as `Update` [completes without error].
> ...
> It isn't quite fair of me to post these runs without also providing exact commits. I can back to this with an update when my commits have settled down.

My conclusion is that `Update` is correct.  Practically, the current branch has the same problem as I had before when I changed `Update` to `Include`.  Theoretically, I think `Update` is correct here because every F# project implicitly depends on `FSharp.Core`, so if we want to adjust that dependency, then we need to use `Update`.

bender2k14 commented:
I just when through this PR and commented on all the things that I thought might be outstanding.  I think it is possible that all of them are already resolved.

When I removed the upper bound on the version FShparp.Core, I did that in a separate commit.  Personally, I would prefer squash that into the commit with message "removed paket", but if you prefer how the branch is now, then I can also leave it as is.

moodmosaic commented:
>I would prefer squash that into the commit with message "removed paket"

I think it's fine keeping it as-is, as a reference. :+1: 

The outcome of `dotnet pack` will be identical (and better!) once I resolve https://github.com/hedgehogqa/fsharp-hedgehog/pull/207#discussion_r452168119 and then we're good to merge this.

bender2k14 commented:
I pushed three more commits.

Phillip Carter help lead me in the right direction.  As I described in detail in https://github.com/dotnet/fsharp/issues/9653#issuecomment-656293313, the issue with the failing tests was caused a previously fixed bug that I accidentally reintroduced in this branch when switching from paket to `<PackageReference />` by not explicitly depending on version `15.8.0` or higher of `Microsoft.NET.Test.SDK`.  I pushed a commit that adds a reference to version `15.9.0` just like the [paket.lock file was doing](https://github.com/hedgehogqa/fsharp-hedgehog/blob/2a68ee22e85b5d9026b3b21a6a0a04bce9cef086/paket.lock#L536).  Then I reverted the commit that moved the `genName` and `genExp` into the test.  I also added a commit with comments about the need to depend on version `15.8.0` or higher of `Microsoft.NET.Test.SDK`.

Again, I would normally squash some of these commits together, but they can also be left as is.

moodmosaic commented:
Cool stuff :100: Thanks a lot for your time and help on this one, it's greatly appreciated :+1:

Microsoft.NET.Test.SDK didn't cross my mind, though I was suspecting that it could be xUnit-related in https://github.com/hedgehogqa/fsharp-hedgehog/pull/208#issuecomment-655409048.

---

I 'll be push a commit to your branch to address the `files` field discussed above, which is basically just following what's described in https://fable.io/docs/your-fable-project/author-a-fable-library.html

(Then feel free to squash any of your commits, if you really want to.)

bender2k14 commented:
> Thanks a lot for your time and help on this one, it's greatly appreciated 👍

You're welcome.  I am glad to help :)

> ...feel free to squash any of your commits, if you really want to.
I do and just did :D

I think commits made during development and PR review should be different than commits that live forever.  I think it is more helpful to have a simpler `master` branch containing the "correct solution".  Anyone that really wants to see all my mistakes and alternative solutions can come to this PR.

I am ready to merge now.

bender2k14 commented:
> Microsoft.NET.Test.SDK didn't cross my mind, though I was suspecting that it could be xUnit-related in [#208 (comment)](https://github.com/hedgehogqa/fsharp-hedgehog/pull/208#issuecomment-655409048).

Ah.  Indeed you did.  Sorry for not giving that comment the full attention it deserved.  I could have save myself some time if I had done so.

moodmosaic commented:
No worries :+1:

moodmosaic commented:
>Anyone that really wants to see all my mistakes and alternative solutions can come to this

So we can then press the squash and merge button on GitHub.

bender2k14 commented:
That is certainly one approach.  My preference is to retain individual commits just squash the "mistakes".  In my project at work, I also prefer to merge by having Azure DevOps rebase and add a merge commit (what it calls a semi-linear merge).

moodmosaic commented:
Alright, do that on your branch and force-push then?

bender2k14 commented:
I squashed and force pushed about 5 hours ago.  I am ready to merge.

moodmosaic commented:
Thanks a lot for all the help, @bender2k14 :100:

moodmosaic commented:
@bender2k14, FWIW, I hope I didn't scare you with all my comments (and nitpicks here and there). I'm pretty sure other OSS projects make it a whole lot easier to get a PR merged...

bender2k14 commented:
Not at all.  I thought all your comments were worth making and the PR was better because of them.  Thanks for being such a great maintainer of this wonderful project!

moodmosaic commented:
Thank you! Well, it may be a wonderful project because there're great contributors like you :rocket:
